restic forget --keep-* options will interpret -1 as "forever"

thndrbrrr opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

`restic forget` will interpret `-1` as "forever" when used with `--keep-*` options. Using a value less than `-1` will result in an error.

~~**Note:** This PR does _not_ change the existing behaviour of `--keep-within*` options until it's decided how durations containing negative values should be handled (see my comment in #2565).~~

`--keep-within*` options containing negative values are considered invalid.

Documentation has not been updated (waiting on confirmation for this PR to go ahead).

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Closes #2565 ~~(if `--keep-within-*` options could be considered a different issue)~~


Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

thndrbrrr committed:
snapshot_group: Fix typo

thndrbrrr committed:
forget: Verify forget opts

thndrbrrr committed:
forget: Treat -1 as forever for all "last n" opts

thndrbrrr committed:
forget: Add test for "keep all hourly snapshots"

thndrbrrr committed:
forget: Add test with regular and "forever" opts

thndrbrrr committed:
forget: Allow neg. values in "--keep-within*" opts

MichaelEischer commented on internal/restic/snapshot_policy.go:
```diff
@@ -260,13 +260,16 @@ func ApplyPolicy(list Snapshots, p ExpirePolicy) (keep, remove Snapshots, reason
 		// Now update the other buckets and see if they have some counts left.
 		for i, b := range buckets {
-			if b.Count > 0 {
+			if b.Count <= -1 || b.Count > 0 {
```
`b.Count == -1` should also be enough? (The current condition is just a complicated way of saying `b.Count != 0`.) And please add a comment, that -1 means infinite.

MichaelEischer commented on internal/restic/snapshot_policy_test.go:
```diff
@@ -251,9 +255,11 @@ func TestApplyPolicy(t *testing.T) {
-				t.Errorf("not enough snapshots removed: policy allows %v snapshots to remain, but ended up with %v",
-					p.Sum(), len(keep))
+			if p.Last >= 0 && p.Hourly >= 0 && p.Daily >= 0 && p.Weekly >= 0 && p.Monthly >= 0 && p.Yearly >= 0 {
+				if p.Sum() > 0 && len(keep) > p.Sum() {
```
`ExpirePolicy.Sum()` is only used by the tests. Thus, please convert that method into a function and move it into the tests. Afterwards, change `Sum()` to return 0 if any of the counts is `-1`.

MichaelEischer commented on cmd/restic/cmd_forget_test.go:
```diff
@@ -0,0 +1,25 @@
+		{Daily: -2},
+		{Weekly: -2},
+		{Monthly: -2},
+		{Yearly: -2},
```
Please also add a basic test to check that -1 and positive counts are accepted.

thndrbrrr replied:
Added tests for -1, 0, and positive values, as well as tests for valid and invalid `Within*` options.

MichaelEischer reviewed (changes requested):
Thanks for working on this issue! The code looks solid except for the comments below. Regarding `--keep-within-*`, I think restic should just return an error if any of them contains a negative duration. I don't see a use case in which that would make sense.

thndrbrrr commented:
Thanks for the feedback, @MichaelEischer! I'll update the PR with the recommended changes. 👍

thndrbrrr commented on cmd/restic/cmd_forget.go:
```diff
@@ -99,8 +99,28 @@ func init() {
+
+	if opts.Last < -1 || opts.Hourly < -1 || opts.Daily < -1 || opts.Weekly < -1 ||
+		opts.Monthly < -1 || opts.Yearly < -1 {
+		negValFound = true
```
Removed `negValFound` and added checks for `--keep-within*` flags

thndrbrrr committed:
Add helper function for Duration parsing

Tests in cmd_forget_test.go need the same convenience function
that was implemented in snapshot_policy_test.go.

Function parseDuration(...) was moved to testing.go  and renamed to
ParseDurationOrPanic(...).

thndrbrrr committed:
forget: Prevent neg. values in --keep-within* opts

thndrbrrr committed:
Refactor policy sum calculation & duration parsing

- Convert policy sum calculation to function and move to tests.
- Remove parseDuration(...) and use ParseDurationOrPanic(...) instead.

thndrbrrr commented:
@MichaelEischer Alrighty, made the changes. This resulted in minor refactoring: had to move (and slightly change) a helper function from `cmd_snapshots_test.go` into `testing.go`.

thndrbrrr commented:
I'll update this PR further with changes to `func (e ExpirePolicy) String() (s string)` to make command line output clearer. Right  now an option like `--keep-daily -1` prints `Applying Policy: keep` ... which misses an important point ... 😉

thndrbrrr committed:
Represent -1 as "all" in ExpirePolicy's Stringer

thndrbrrr commented:
Ok, updated PR. Options such as `--keep-daily 7 --keep-yearly -1 --keep-monthly -1 --keep-hourly 24` for example will now print:
```
Applying Policy: keep 24 hourly, 7 daily, all monthly, all yearly snapshots
```

thndrbrrr commented:
@MichaelEischer Any feedback on the updated PR? Happy to update the docs if that's all that's misisng at this point.

thndrbrrr committed:
Fix comment to comply with linter

thndrbrrr committed:
Add changelog entry

thndrbrrr committed:
forget: Update documentation

thndrbrrr committed:
forget: Update usage text

thndrbrrr commented on internal/restic/testing.go:
```diff
@@ -212,3 +212,14 @@ func TestParseHandle(s string, t BlobType) BlobHandle {
 	sn.id = &id
 }
+
+// Convenience function that parses a duration from a string or panics if string is invalid.
```
Done.

thndrbrrr commented:
@MichaelEischer Docs and changelog updated.

I updated the usage text as well, let me know if you're not happy with the wording:

```
$ ./restic forget -h
[...]
Usage:
  restic forget [flags] [snapshot ID] [...]

Flags:
  -l, --keep-last n                    keep the last n snapshots (use "-1" for n to keep all snapshots)
  -H, --keep-hourly n                  keep the last n hourly snapshots (use "-1" for n to keep all hourly snapshots)
  -d, --keep-daily n                   keep the last n daily snapshots (use "-1" for n to keep all daily snapshots)
  -w, --keep-weekly n                  keep the last n weekly snapshots (use "-1" for n to keep all weekly snapshots)
  -m, --keep-monthly n                 keep the last n monthly snapshots (use "-1" for n to keep all monthly snapshots)
  -y, --keep-yearly n                  keep the last n yearly snapshots (use "-1" for n to keep all yearly snapshots)
```

thndrbrrr commented:
> @thndrbrrr Sorry for the late reply, I've been far too busy during the last few weeks.

No worries, I know how that feels ... 😆

MichaelEischer commented on cmd/restic/cmd_forget.go:
```diff
@@ -67,12 +67,12 @@ func init() {
-	f.IntVarP(&forgetOptions.Weekly, "keep-weekly", "w", 0, "keep the last `n` weekly snapshots")
-	f.IntVarP(&forgetOptions.Monthly, "keep-monthly", "m", 0, "keep the last `n` monthly snapshots")
-	f.IntVarP(&forgetOptions.Yearly, "keep-yearly", "y", 0, "keep the last `n` yearly snapshots")
+	f.IntVarP(&forgetOptions.Last, "keep-last", "l", 0, "keep the last `n` snapshots (use \"-1\" for n to keep all snapshots)")
```
My feeling is that `(use \"-1\" to keep all snapshots)` might be easier to understand. After all, there's no other place in which one could add the `-1`.

thndrbrrr replied:
👍

MichaelEischer reviewed (approved):
I only have one small remaining nitpick. Besides that LGTM.

thndrbrrr committed:
forget: Simplify usage text
